projects/fluid_simulation/sim.cpp

```diff
@@ -0,0 +1,435 @@
+    PPX_CHECKED_CALL(GetApp()->GetDevice()->CreateFence(&fci, &frame.imageAcquiredFence));
+    PPX_CHECKED_CALL(GetApp()->GetDevice()->CreateSemaphore(&sci, &frame.renderCompleteSemaphore));
+    fci = {true}; // Create signaled
+    PPX_CHECKED_CALL(GetApp()->GetDevice()->CreateFence(&fci, &frame.renderCompleteFence));
+    mPerFrame.push_back(frame);
+
+    // Set up all the filters to use.
+    InitComputeShaders();
+
+    // Set up draw program to emit computed textures to the swapchain.
+    InitGraphicShaders();
+
+    // Initialize textures
+    InitTextures();
+}
+
+void FluidSimulation::InitComputeShaders()
+{
+    // Descriptor set layout.  This must match assets/fluid_simulation/shaders/config.hlsli and it
+    // is shared across all ComputeShader instances.
+    ppx::grfx::DescriptorSetLayoutCreateInfo lci = {};
+    lci.bindings.push_back(ppx::grfx::DescriptorBinding(0, ppx::grfx::DESCRIPTOR_TYPE_UNIFORM_BUFFER));
+    lci.bindings.push_back(ppx::grfx::DescriptorBinding(1, ppx::grfx::DESCRIPTOR_TYPE_SAMPLER));
+    lci.bindings.push_back(ppx::grfx::DescriptorBinding(2, ppx::grfx::DESCRIPTOR_TYPE_SAMPLED_IMAGE));
+    lci.bindings.push_back(ppx::grfx::DescriptorBinding(3, ppx::grfx::DESCRIPTOR_TYPE_SAMPLED_IMAGE));
+    lci.bindings.push_back(ppx::grfx::DescriptorBinding(4, ppx::grfx::DESCRIPTOR_TYPE_SAMPLED_IMAGE));
+    lci.bindings.push_back(ppx::grfx::DescriptorBinding(5, ppx::grfx::DESCRIPTOR_TYPE_SAMPLED_IMAGE));
+    lci.bindings.push_back(ppx::grfx::DescriptorBinding(6, ppx::grfx::DESCRIPTOR_TYPE_SAMPLED_IMAGE));
+    lci.bindings.push_back(ppx::grfx::DescriptorBinding(7, ppx::grfx::DESCRIPTOR_TYPE_SAMPLED_IMAGE));
+    lci.bindings.push_back(ppx::grfx::DescriptorBinding(8, ppx::grfx::DESCRIPTOR_TYPE_SAMPLED_IMAGE));
+    lci.bindings.push_back(ppx::grfx::DescriptorBinding(9, ppx::grfx::DESCRIPTOR_TYPE_SAMPLED_IMAGE));
+    lci.bindings.push_back(ppx::grfx::DescriptorBinding(10, ppx::grfx::DESCRIPTOR_TYPE_SAMPLED_IMAGE));
+    lci.bindings.push_back(ppx::grfx::DescriptorBinding(11, ppx::grfx::DESCRIPTOR_TYPE_STORAGE_IMAGE));
+    PPX_CHECKED_CALL(GetApp()->GetDevice()->CreateDescriptorSetLayout(&lci, &mCompute.mDescriptorSetLayout));
+
+    // Compute pipeline interface.
+    ppx::grfx::PipelineInterfaceCreateInfo pici = {};
+    pici.setCount                               = 1;
+    pici.sets[0].set                            = 0;
+    pici.sets[0].pLayout                        = mCompute.mDescriptorSetLayout;
+    PPX_CHECKED_CALL(GetApp()->GetDevice()->CreatePipelineInterface(&pici, &mCompute.mPipelineInterface));
+
+    // Compute sampler.
+    ppx::grfx::SamplerCreateInfo sci = {};
+    sci.magFilter                    = ppx::grfx::FILTER_LINEAR;
+    sci.minFilter                    = ppx::grfx::FILTER_LINEAR;
+    sci.mipmapMode                   = ppx::grfx::SAMPLER_MIPMAP_MODE_NEAREST;
+    sci.addressModeU                 = ppx::grfx::SAMPLER_ADDRESS_MODE_CLAMP_TO_EDGE;
```

Comment:
Sorry.  Can't see the difference here :)

Reply:
The difference is pretty stark,  I suggest applying the fix suggested by Nathan

Reply:
Your comment doesn't actually help me see the difference.  Can you describe it?

Reply:
There are very visible bands according to the arrows in Nathan's pic. Is your monitor uncalibrated maybe (it's dark, but very visible for me on both phone and PC monitor). I cropped it more, can you see this?
![image](https://user-images.githubusercontent.com/3148446/216679291-a67a72a9-70fe-4b38-9ce5-4ac9332b1204.png)

Reply:
Oh! The banding on the top corner?  OK, I see it now.  Not sure how you generated that image, Nathan.  The output I'm getting on my machine is completely different.

![image](https://user-images.githubusercontent.com/7295335/216679497-4a7ce35b-9946-4f4a-8405-da7b78340eec.png)

Reply:
It's just slightly visible on the full-color image (maybe pushing gamma or contract would show it?)
But I found out by playing around. I'll make a list of items to remove from the shader to reproduce it in a visible way.

Reply:
It's actually visible on your screenshot. lighter than on my repro, but visible (look on the top-left corner, and follow the horizontal or vertical lines, as seen in Daniele's zooms)
![visible](https://user-images.githubusercontent.com/4282252/216927584-418b4b2a-4dcd-4219-8869-54179f413a1c.png)

Reply:
I applied it to all the shaders.  It's the simplest approach atm since all the compute shaders share this code.  I will continue playing with it, but the results I got now are not awful (see attached), so I'll keep it simple.

![Screenshot 2023-03-24 145616](https://user-images.githubusercontent.com/7295335/227615507-a3c0fb2e-0428-47b8-86f0-2ce525629afe.png)

Reply:
Filed #51 to follow-up later.